### PR TITLE
38947_added_log_in_log_out

### PR DESCRIPTION
**Pull request by hasanmainul**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      redirect_to user
+    else
+      flash.now[:danger] = "Invalid email/password combination"
```
Comment (chuanhtuan) [inline app/controllers/sessions_controller.rb]: i18n

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
```
Comment (chuanhtuan) [inline app/helpers/sessions_helper.rb]: please follow by tutorial
(have to implement remember me)
:(

File: app/models/user.rb [hunk 3]
```diff
@@ -7,4 +8,28 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (chuanhtuan) [inline app/models/user.rb]: use update_attributes instead of update_attribute

File: app/models/user.rb [hunk 4]
```diff
@@ -7,4 +8,28 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new remember_digest.is_password?(remember_token)
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (chuanhtuan) [inline app/models/user.rb]: use update_attributes instead of update_attribute

File: app/models/user.rb [hunk 5]
```diff
@@ -7,4 +8,28 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attributes :remember_digest => User.digest(remember_token)
```
Comment (chuanhtuan) [inline app/models/user.rb]: update_attributes remember_digest: User.digest(remember_token)

Comment (chuanhtuan) [inline app/models/user.rb]: please try with update_attributes!, you will get a error :D

File: app/models/user.rb [hunk 6]
```diff
@@ -7,4 +8,28 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new remember_digest.is_password?(remember_token)
+  end
+
+  def forget
+    update_attributes :remember_digest => nil
```
Comment (chuanhtuan) [inline app/models/user.rb]: update_attributes remember_digest: nil

File: app/models/user.rb [hunk 7]
```diff
@@ -7,4 +8,28 @@ class User < ActiveRecord::Base
...
+    BCrypt::Password.create(string, cost: cost)
+  end
+
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attributes remember_digest: User.digest(remember_token)
```
Comment (chuanhtuan) [inline app/models/user.rb]: try with update_attributes!, you will get a error, please fix it
update_attributes! remember_digest: User.digest(remember_token)

File: app/models/user.rb [hunk 8]
```diff
@@ -7,4 +8,28 @@ class User < ActiveRecord::Base
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  def forget
+    update_attributes remember_digest: nil
```
Comment (chuanhtuan) [inline app/models/user.rb]: try with update_attributes!, you will get a error, please fix it
update_attributes! remember_digest: nil

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
```
Comment (dieunb) [inline app/controllers/sessions_controller.rb]: use double quotation marks instead of single quotation marks

**Pull request merged**